pkg/gatewayserver/io/rtts.go

```diff
@@ -0,0 +1,74 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package io
+
+import (
+	"sort"
+	"sync"
+	"time"
+)
+
+type rtts struct {
+	count int
+	mu    sync.RWMutex
+	items []time.Duration
+}
+
+func newRTTs(count int) *rtts {
+	return &rtts{
+		count: count,
+		items: make([]time.Duration, 0, count+1),
+	}
+}
+
+// Record records the given round-trip time.
+func (r *rtts) Record(d time.Duration) {
+	r.mu.Lock()
+	r.items = append(r.items, d)
+	if len(r.items) > r.count {
+		r.items = append(r.items[:0], r.items[len(r.items)-r.count:]...)
```

Comment:
This means that we're going to copy the entire slice every time `Record` is called after it reaches `count`. Why not just reslice with `r.items = r.items[len(r.items)-r.count:]`? The call to `Stats` already makes a copy before it sorts, so that isn't an issue.

Reply:
That creates a memory leak, from [Go Slices: usage and internals](https://blog.golang.org/go-slices-usage-and-internals):

>Slicing does not copy the slice's data. It creates a new slice value that points to the original array. This makes slice operations as efficient as manipulating array indices.

>As mentioned earlier, re-slicing a slice doesn't make a copy of the underlying array. The full array will be kept in memory until it is no longer referenced. Occasionally this can cause the program to hold all the data in memory when only a small piece of it is needed.

That "small piece" is your sliding tail. At least, that's what I understand; it's not that the array is cut when there's no more reference to earlier items in the array.